### PR TITLE
fix: deep assign should ignore inseparable object

### PR DESCRIPTION
**Pull request by haoxins**

### CODE REVIEW

**Comment by haoxins** [comment]
also fix failed test: support `Object.create(null)` targets

**Comment by schnittstabil** [comment]
Thank you @coderhaoxin, but that's not how Object.assign works, e.g:

``` js
// this PR:
> deepAssign({regexp: /reg/}, {regexp: 'string'});
{regexp: 'string'}

// but:
> Object.assign(/reg/, 'string')
{ /reg/ '0': 's', '1': 't', '2': 'r', '3': 'i', '4': 'n', '5': 'g' }
```

File: test.js [hunk 1]
```diff
@@ -137,7 +165,7 @@ test('support `Object.create(null)` objects', t => {
 test('support `Object.create(null)` targets', t => {
 	var obj = Object.create(null);
 	obj.foo = true;
-	t.same(fn(obj, {bar: false}), {foo: true, bar: false});
+	t.same(fn({}, obj, {bar: false}), {foo: true, bar: false});
```
Comment (schnittstabil) [inline test.js]: Nothing is wrong with this test. The first deep-assign argument is the target and this test asserts, that the following does not throw an Error:

``` js
deep-assign(Object.create(null), …)
```

**Comment by coderhaoxin** [comment]
@schnittstabil below is my test on Chrome48 (It seems that v8 changed something)

```js
Object.assign(/reg/, 'string')
/reg/
Object.assign({name: /reg/}, {name: 'string'})
Object {name: "string"}
```

node (v5)

```js
Object.assign(/reg/, 'string')
{ /reg/ '0': 's', '1': 't', '2': 'r', '3': 'i', '4': 'n', '5': 'g' }

Object.assign({name: /reg/}, {name: 'string'})
{ name: 'string' }
```

**Comment by coderhaoxin** [comment]
also

```js
Object.assign({date: new Date()}, {date: '123'})
Object {date: "123"}

Object.assign({date: new Date()}, {date: new Date(1445428088488)})
Object {date: Wed Oct 21 2015 19:48:08 GMT+0800 (CST)}
```

File: test.js (hunk 1 above)
Comment (coderhaoxin) [inline test.js]: OK, I'll revert this

**Comment by haoxins** [comment]
@schnittstabil  Oh, I miss something. Thx

**Comment by haoxins** [comment]
@schnittstabil `Object.assign()` convert all the arguments to `Object`, but should `deep-assign` the same?

I mean, is it reasonable that 

``` js
deepAssign({regexp: /reg/}, {regexp: 'string'});
```

should equal to

``` js
Object.assign(/reg/, 'string')
```

for the property `regexp`?

**Comment by schnittstabil** [comment]
The project description of deep-assign says:

> Recursive Object.assign()

@coderhaoxin Therefore I would say yes, in my opinion it is the canonical way.

But I must admit that the behavior is not very obvious, which is shown by #3, #4, #6. I am open to new interpretations of _recursive_ and _assign_, but `deep-assign` should behave like `Object.assign` in the base case…

File: test.js (hunk 1 above)
Comment (schnittstabil) [inline test.js]: @coderhaoxin Thank you for mentioning that. You were right, this test had to be repaired (e252fa2dec757a6c608e242ca49ca060f2439ad2).

**Comment by schnittstabil** [comment]
@coderhaoxin [merge-options](https://github.com/schnittstabil/merge-options) might be of interest to you.

**Pull request closed without merging**